### PR TITLE
[#1074] security policy support separate authentication token and api

### PR DESCRIPTION
**Pull request by chengyouling**

### CODE REVIEW

File: spring-cloud-huawei-governance/src/main/java/com/huaweicloud/governance/authentication/RSAProviderTokenManager.java [hunk 1]
```diff
@@ -49,57 +34,12 @@ public RSAProviderTokenManager(List<AccessController> accessControllers) {
...
-          validatedToken.put(rsaToken, true);
-          isAllow = accessController.isAllowed(rsaToken.getServiceId(), rsaToken.getInstanceId(), requestMap);
-        }
-        if (!isAllow) {
-          throw new UnAuthorizedException(accessController.interceptMessage());
-        }
+        accessController.valid(token, requestMap);
       }
-    } catch (InvalidKeyException | NoSuchAlgorithmException | InvalidKeySpecException | SignatureException e) {
+    }catch(Exception e){
```
Comment (liubao68) [inline spring-cloud-huawei-governance/src/main/java/com/huaweicloud/governance/authentication/RSAProviderTokenManager.java:40]: format code

File: spring-cloud-huawei-governance/src/main/java/com/huaweicloud/governance/authentication/securityPolicy/SecurityPolicyAccessController.java [hunk 2]
```diff
@@ -38,17 +44,41 @@ public SecurityPolicyAccessController(AuthenticationAdapter authenticationAdapte
   }
 
   @Override
-  public boolean isAllowed(String serviceId, String instanceId, Map<String, String> requestMap) {
+  public void valid(String token, Map<String, String> requestMap) throws Exception {
+    RsaAuthenticationToken rsaToken = checkTokenInfoOrServiceName(token);
```
Comment (liubao68) [inline spring-cloud-huawei-governance/src/main/java/com/huaweicloud/governance/authentication/securityPolicy/SecurityPolicyAccessController.java:48]: when token check is enabled while is not valid, should fail.

File: integration-tests/discovery-servicecomb-tests/account-provider/src/main/resources/bootstrap.yaml [hunk 3]
```diff
@@ -59,6 +59,7 @@ spring:
 
           securityPolicyEnabled: true
           acls:
+            tokenCheckEnabled: true
```
Comment (liubao68) [inline integration-tests/discovery-servicecomb-tests/account-provider/src/main/resources/bootstrap.yaml:62]: add as a global configuration not only for acls.

File: spring-cloud-huawei-governance/src/main/java/com/huaweicloud/governance/authentication/RSAProviderTokenManager.java [hunk 4]
```diff
@@ -49,57 +34,12 @@ public RSAProviderTokenManager(List<AccessController> accessControllers) {
...
-        if (validatedToken.asMap().containsKey(rsaToken)) {
-          isAllow = accessController.isAllowed(rsaToken.getServiceId(), rsaToken.getInstanceId(), requestMap);
-        } else if (isValidToken(rsaToken, accessController) && !tokenExpired(rsaToken)) {
-          validatedToken.put(rsaToken, true);
-          isAllow = accessController.isAllowed(rsaToken.getServiceId(), rsaToken.getInstanceId(), requestMap);
-        }
-        if (!isAllow) {
-          throw new UnAuthorizedException(accessController.interceptMessage());
-        }
+        accessController.valid(token, requestMap);
```
Comment (liubao68) [inline spring-cloud-huawei-governance/src/main/java/com/huaweicloud/governance/authentication/RSAProviderTokenManager.java:38]: Check token first before call accessController.valid.

File: spring-cloud-huawei-governance/src/main/java/com/huaweicloud/governance/authentication/RSAProviderTokenManager.java (hunk 1 above)
Comment (chengyouling) [reply spring-cloud-huawei-governance/src/main/java/com/huaweicloud/governance/authentication/RSAProviderTokenManager.java]: fixed

File: spring-cloud-huawei-governance/src/main/java/com/huaweicloud/governance/authentication/securityPolicy/SecurityPolicyAccessController.java (hunk 2 above)
Comment (chengyouling) [reply spring-cloud-huawei-governance/src/main/java/com/huaweicloud/governance/authentication/securityPolicy/SecurityPolicyAccessController.java]: fixed

File: integration-tests/discovery-servicecomb-tests/account-provider/src/main/resources/bootstrap.yaml (hunk 3 above)
Comment (chengyouling) [reply integration-tests/discovery-servicecomb-tests/account-provider/src/main/resources/bootstrap.yaml]: fixed

File: spring-cloud-huawei-governance/src/main/java/com/huaweicloud/governance/authentication/RSAProviderTokenManager.java (hunk 4 above)
Comment (chengyouling) [reply spring-cloud-huawei-governance/src/main/java/com/huaweicloud/governance/authentication/RSAProviderTokenManager.java]: fixed

File: spring-cloud-huawei-governance/src/main/java/com/huaweicloud/governance/authentication/securityPolicy/SecurityPolicyAccessController.java [hunk 5]
```diff
@@ -16,39 +16,67 @@
...
     this.authenticationAdapter = authenticationAdapter;
     this.securityPolicyProperties = securityPolicyProperties;
+    this.environment = environment;
   }
 
   @Override
-  public boolean isAllowed(String serviceId, String instanceId, Map<String, String> requestMap) {
-    return checkAllow(serviceId, requestMap) && !checkDeny(serviceId, requestMap);
+  public RsaAuthenticationToken validProcess(String token, String serviceName) throws Exception {
+    if (Boolean.parseBoolean(environment.getProperty(Const.AUTH_TOKEN_CHECK_ENABLED, "true"))) {
```
Comment (liubao68) [inline spring-cloud-huawei-governance/src/main/java/com/huaweicloud/governance/authentication/securityPolicy/SecurityPolicyAccessController.java:52]: `environment.getProperty(.., boolean.class, true) `

File: integration-tests/discovery-servicecomb-tests/account-provider/src/main/resources/bootstrap.yaml [hunk 6]
```diff
@@ -38,6 +38,7 @@ spring:
         firstPullRequired: false
 
       webmvc:
+        tokenCheckEnabled: true
```
Comment (liubao68) [inline integration-tests/discovery-servicecomb-tests/account-provider/src/main/resources/bootstrap.yaml:41]: Under publicKey

File: spring-cloud-huawei-governance/src/main/java/com/huaweicloud/governance/authentication/RSATokenCheckUtils.java [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+
+public class RSATokenCheckUtils {
+  private static final Logger LOGGER = LoggerFactory.getLogger(RSATokenCheckUtils.class);
+
+  public static final Cache<RsaAuthenticationToken, Boolean> validatedToken = CacheBuilder.newBuilder()
```
Comment (liubao68) [inline spring-cloud-huawei-governance/src/main/java/com/huaweicloud/governance/authentication/RSATokenCheckUtils.java:32]: field never write

File: spring-cloud-huawei-governance/src/main/java/com/huaweicloud/governance/authentication/AccessController.java [hunk 8]
```diff
@@ -20,11 +20,9 @@
  * Add black / white list control to service access
  */
 public interface AccessController {
-  RsaAuthenticationToken validProcess(String token, String serviceName) throws Exception;
-
   String getPublicKeyFromInstance(String instanceId, String serviceId);
 
-  boolean isAllowed(Map<String, String> requestMap);
+  boolean isAllowed(Map<String, String> requestMap, boolean isNeedCheckServiceName) throws Exception;
```
Comment (liubao68) [inline spring-cloud-huawei-governance/src/main/java/com/huaweicloud/governance/authentication/AccessController.java:25]:
```
boolean isAllowed(RsaAuthenticationToken token, String serviceName)
```

File: spring-cloud-huawei-governance/src/main/java/com/huaweicloud/governance/authentication/RSAProviderTokenManager.java [hunk 9]
```diff
@@ -21,33 +21,41 @@
...
     this.accessControllers = accessControllers;
+    this.environment = environment;
   }
 
   public void valid(String token, Map<String, String> requestMap) throws Exception {
     try {
-      boolean isAllow = true;
+      RsaAuthenticationToken rsaToken = null;
+      if (environment.getProperty(Const.AUTH_TOKEN_CHECK_ENABLED, Boolean.class, true)) {
+        rsaToken = RSATokenCheckUtils.checkTokenInfo(token);
```
Comment (liubao68) [inline spring-cloud-huawei-governance/src/main/java/com/huaweicloud/governance/authentication/RSAProviderTokenManager.java:42]: Hide `RSATokenCheckUtils.validatedToken.put(rsaToken, true);` in  `RSATokenCheckUtils.checkTokenInfo(token)`

File: spring-cloud-huawei-governance/src/main/java/com/huaweicloud/governance/authentication/RSAProviderTokenManager.java [hunk 10]
```diff
@@ -21,33 +21,41 @@
...
+  public RSAProviderTokenManager(List<AccessController> accessControllers, Environment environment) {
     this.accessControllers = accessControllers;
+    this.environment = environment;
   }
 
   public void valid(String token, Map<String, String> requestMap) throws Exception {
     try {
-      boolean isAllow = true;
+      RsaAuthenticationToken rsaToken = null;
+      if (environment.getProperty(Const.AUTH_TOKEN_CHECK_ENABLED, Boolean.class, true)) {
```
Comment (liubao68) [inline spring-cloud-huawei-governance/src/main/java/com/huaweicloud/governance/authentication/RSAProviderTokenManager.java:41]: `environment.getProperty(Const.AUTH_TOKEN_CHECK_ENABLED, boolean.class, true)`

File: integration-tests/discovery-servicecomb-tests/account-provider/src/main/resources/bootstrap.yaml (hunk 6 above)
Comment (chengyouling) [reply integration-tests/discovery-servicecomb-tests/account-provider/src/main/resources/bootstrap.yaml]: fixed

File: spring-cloud-huawei-governance/src/main/java/com/huaweicloud/governance/authentication/RSATokenCheckUtils.java (hunk 7 above)
Comment (chengyouling) [reply spring-cloud-huawei-governance/src/main/java/com/huaweicloud/governance/authentication/RSATokenCheckUtils.java:32]: fixed

File: spring-cloud-huawei-governance/src/main/java/com/huaweicloud/governance/authentication/AccessController.java (hunk 8 above)
Comment (chengyouling) [reply spring-cloud-huawei-governance/src/main/java/com/huaweicloud/governance/authentication/AccessController.java]: fixed

File: spring-cloud-huawei-governance/src/main/java/com/huaweicloud/governance/authentication/RSAProviderTokenManager.java (hunk 9 above)
Comment (chengyouling) [reply spring-cloud-huawei-governance/src/main/java/com/huaweicloud/governance/authentication/RSAProviderTokenManager.java:43]: fixed

File: spring-cloud-huawei-governance/src/main/java/com/huaweicloud/governance/authentication/RSAProviderTokenManager.java (hunk 10 above)
Comment (chengyouling) [reply spring-cloud-huawei-governance/src/main/java/com/huaweicloud/governance/authentication/RSAProviderTokenManager.java]: fixed

File: spring-cloud-huawei-governance/src/main/java/com/huaweicloud/governance/authentication/ProviderAuthPreHandlerInterceptor.java [hunk 11]
```diff
@@ -31,18 +34,21 @@ public class ProviderAuthPreHandlerInterceptor implements PreHandlerInterceptor
...
+    if (StringUtils.isEmpty(serviceName)) {
+      serviceName = InvocationContextHolder.getOrCreateInvocationContext().getContext(Const.AUTH_SERVICE_NAME);
+    }
     Map<String, String> requestMap = new HashMap<>();
-    requestMap.put("uri", request.getRequestURI());
-    requestMap.put("method", request.getMethod());
+    requestMap.put(Const.AUTH_URI, request.getRequestURI());
+    requestMap.put(Const.AUTH_METHOD, request.getMethod());
+    requestMap.put(Const.AUTH_SERVICE_NAME, serviceName);
     authenticationTokenManager.valid(token, requestMap);
```
Comment (liubao68) [inline spring-cloud-huawei-governance/src/main/java/com/huaweicloud/governance/authentication/ProviderAuthPreHandlerInterceptor.java:52]: 参考 GovernanceExtractor 设计，不要将数据存储在临时 MAP 里面， 会降低性能。

File: spring-cloud-huawei-governance/src/main/java/com/huaweicloud/governance/authentication/AccessController.java [hunk 12]
```diff
@@ -20,10 +20,9 @@
  * Add black / white list control to service access
  */
 public interface AccessController {
-
-  boolean isAllowed(String serviceId, String instanceId, Map<String, String> requestMap);
-
   String getPublicKeyFromInstance(String instanceId, String serviceId);
 
+  boolean isAllowed(Map<String, String> requestMap, String serviceName) throws Exception;
```
Comment (liubao68) [inline spring-cloud-huawei-governance/src/main/java/com/huaweicloud/governance/authentication/AccessController.java:25]: `boolean isAllowed(AuthRequestExtractor request)`

File: spring-cloud-huawei-governance/src/main/java/com/huaweicloud/governance/authentication/securityPolicy/SecurityPolicyAccessController.java [hunk 13]
```diff
@@ -38,17 +42,31 @@ public SecurityPolicyAccessController(AuthenticationAdapter authenticationAdapte
...
+  private boolean checkDeny(String serviceName, Map<String, String> requestMap) {
+    if (securityPolicyProperties.matchDeny(serviceName, requestMap.get(Const.AUTH_URI),
+        requestMap.get(Const.AUTH_METHOD))) {
       // permissive mode, black policy match allow passing
       if ("permissive".equals(securityPolicyProperties.getMode())) {
-        LOGGER.info("[autoauthz unauthorized request] consumer={}, provider={}, path={}, method={}, timestamp={}",
-            serviceId, securityPolicyProperties.getProvider(), requestMap.get("method"), requestMap.get("uri"),
-            System.currentTimeMillis());
+        requestMap.put(Const.ALARM_MESSAGE_KEY, buildAlarmMessage(serviceName, requestMap.get(Const.AUTH_URI),
+            requestMap.get(Const.AUTH_METHOD)));
```
Comment (liubao68) [inline spring-cloud-huawei-governance/src/main/java/com/huaweicloud/governance/authentication/securityPolicy/SecurityPolicyAccessController.java:69]: 信息在记录日志的时候创建

File: spring-cloud-huawei-nacos/src/main/java/com/huaweicloud/nacos/authentication/NacosAuthenticationAdapter.java [hunk 14]
```diff
@@ -85,6 +87,15 @@ public String getPropertyValue(String serviceId, String instanceId, String prope
     }
   }
 
+  @Override
+  public String getServiceName(String serviceId) {
+    List<ServiceInstance> instances = serviceDiscovery.getInstances(serviceId);
+    if (!CollectionUtils.isEmpty(instances)) {
+      return instances.get(0).getServiceId();
+    }
+    return serviceId;
```
Comment (liubao68) [inline spring-cloud-huawei-nacos/src/main/java/com/huaweicloud/nacos/authentication/NacosAuthenticationAdapter.java:96]: return serviceId;

File: spring-cloud-huawei-governance/src/main/java/com/huaweicloud/governance/authentication/AccessController.java (hunk 12 above)
Comment (chengyouling) [reply spring-cloud-huawei-governance/src/main/java/com/huaweicloud/governance/authentication/AccessController.java]: fixed

File: spring-cloud-huawei-governance/src/main/java/com/huaweicloud/governance/authentication/securityPolicy/SecurityPolicyAccessController.java (hunk 13 above)
Comment (chengyouling) [reply spring-cloud-huawei-governance/src/main/java/com/huaweicloud/governance/authentication/securityPolicy/SecurityPolicyAccessController.java]: fixed

File: spring-cloud-huawei-nacos/src/main/java/com/huaweicloud/nacos/authentication/NacosAuthenticationAdapter.java (hunk 14 above)
Comment (chengyouling) [reply spring-cloud-huawei-nacos/src/main/java/com/huaweicloud/nacos/authentication/NacosAuthenticationAdapter.java]: fixed

File: spring-cloud-huawei-governance/src/main/java/com/huaweicloud/governance/authentication/ProviderAuthPreHandlerInterceptor.java (hunk 11 above)
Comment (chengyouling) [reply spring-cloud-huawei-governance/src/main/java/com/huaweicloud/governance/authentication/ProviderAuthPreHandlerInterceptor.java]: fixed

File: spring-cloud-huawei-governance/src/main/java/com/huaweicloud/governance/authentication/RSAProviderTokenManager.java [hunk 15]
```diff
@@ -34,17 +36,22 @@ public RSAProviderTokenManager(List<AccessController> accessControllers, Environ
...
+      } else {
+        extractor = AuthRequestExtractorUtils.createAuthRequestExtractor(request, "", "");
       }
       boolean isAllow;
       for (AccessController accessController : accessControllers) {
         if (rsaToken != null) {
-          extractor.setServiceId(rsaToken.getServiceId());
-          extractor.setInstanceId(rsaToken.getInstanceId());
           if (RSATokenCheckUtils.validTokenInfo(rsaToken,
               accessController.getPublicKeyFromInstance(rsaToken.getInstanceId(), rsaToken.getServiceId()))) {
```
Comment (liubao68) [inline spring-cloud-huawei-governance/src/main/java/com/huaweicloud/governance/authentication/RSAProviderTokenManager.java:56]: validTokenInfo相关的检查放到AccessControlleris.Allowed前面， 并且和 checkTokenInfo  合并。

File: spring-cloud-huawei-governance/src/main/java/com/huaweicloud/governance/authentication/RSATokenCheckUtils.java [hunk 16]
```diff
@@ -55,7 +56,15 @@ public static RsaAuthenticationToken checkTokenInfo(HttpServletRequest request)
...
       throw new UnAuthorizedException("UNAUTHORIZED.");
     }
-    validatedToken.put(rsaToken, true);
+    if (validatedToken.asMap().containsKey(rsaToken)) {
+      return rsaToken;
+    }
+    if (isValidToken(rsaToken, authenticationAdapter) && !tokenExpired(rsaToken)) {
+      validatedToken.put(rsaToken, true);
+    } else {
+      LOGGER.error("token is expired, restart service.");
```
Comment (liubao68) [inline spring-cloud-huawei-governance/src/main/java/com/huaweicloud/governance/authentication/RSATokenCheckUtils.java:65]: logs is not correct

File: spring-cloud-huawei-governance/src/main/java/com/huaweicloud/governance/authentication/RSATokenCheckUtils.java [hunk 17]
```diff
@@ -55,7 +56,15 @@ public static RsaAuthenticationToken checkTokenInfo(HttpServletRequest request)
       LOGGER.error("token is expired");
       throw new UnAuthorizedException("UNAUTHORIZED.");
     }
-    validatedToken.put(rsaToken, true);
+    if (validatedToken.asMap().containsKey(rsaToken)) {
+      return rsaToken;
+    }
+    if (isValidToken(rsaToken, authenticationAdapter) && !tokenExpired(rsaToken)) {
```
Comment (liubao68) [inline spring-cloud-huawei-governance/src/main/java/com/huaweicloud/governance/authentication/RSATokenCheckUtils.java:62]: tokenExpired has already checked.

File: spring-cloud-huawei-governance/src/main/java/com/huaweicloud/governance/authentication/RSAProviderTokenManager.java (hunk 15 above)
Comment (chengyouling) [reply spring-cloud-huawei-governance/src/main/java/com/huaweicloud/governance/authentication/RSAProviderTokenManager.java]: fixed

File: spring-cloud-huawei-governance/src/main/java/com/huaweicloud/governance/authentication/RSATokenCheckUtils.java (hunk 17 above)
Comment (chengyouling) [reply spring-cloud-huawei-governance/src/main/java/com/huaweicloud/governance/authentication/RSATokenCheckUtils.java]: fixed

File: spring-cloud-huawei-governance/src/main/java/com/huaweicloud/governance/authentication/RSATokenCheckUtils.java (hunk 16 above)
Comment (chengyouling) [reply spring-cloud-huawei-governance/src/main/java/com/huaweicloud/governance/authentication/RSATokenCheckUtils.java]: fixed

**Review by liubao68 — APPROVED** [review]

**Pull request merged**